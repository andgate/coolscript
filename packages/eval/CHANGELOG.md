# @coolscript/eval

## 0.1.2

### Patch Changes

- 6349a12: Testing changeset workflow

## 0.1.1

### Patch Changes

- da87a04: bump patches for all changesets
