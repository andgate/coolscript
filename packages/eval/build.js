const process = require('process')
const esbuild = require('esbuild')
const { dependencies } = require('./package.json')
const external = Object.keys(dependencies)

let watch = process.argv.some((arg) => arg == '--watch')
if (watch) {
  watch = {
    onRebuild(error, result) {
      if (error) console.error('watch build failed:', error)
      else console.log('watch build succeeded:', result)
    }
  }
}

const shared = {
  entryPoints: ['src/index.ts'],
  outdir: 'lib/',
  target: 'es6',
  sourcemap: true,
  bundle: true,
  minify: true,
  tsconfig: 'tsconfig.build.json',
  external,
  watch
}

esbuild
  .build({
    ...shared,
    format: 'esm',
    splitting: true
  })
  .catch(() => process.exit(1))

esbuild
  .build({
    ...shared,
    format: 'cjs',
    outExtension: { '.js': '.cjs' }
  })
  .catch(() => process.exit(1))
