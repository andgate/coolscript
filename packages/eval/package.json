{
  "name": "@coolscript/eval",
  "version": "0.0.4",
  "description": "Coolscript programming language evaluator.",
  "author": "Gabriel Anderson",
  "license": "MIT",
  "main": "./lib/index.js",
  "module": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "publishConfig": {
    "main": "./lib/index.js",
    "module": "./lib/index.js",
    "typings": "lib/index.d.ts",
    "access": "public"
  },
  "files": [
    "lib/",
    "LICENSE"
  ],
  "scripts": {
    "start": "pnpm watch",
    "build": "pnpm build:lib",
    "build:lib": "pnpm clean && pnpm build:ts && pnpm build:js",
    "build:js": "node build.js",
    "build:ts": "tsc --project tsconfig.build.json",
    "clean": "rimraf lib/ *.tsbuildinfo",
    "watch": "concurrently \"pnpm:watch:*\"",
    "watch:js": "node build.js --watch",
    "watch:ts": "tsc --project tsconfig.build.json --watch",
    "check": "tsc --project tsconfig.build.json --noEmit",
    "prettier": "prettier --check src/ tests/",
    "prettier:fix": "prettier --write src/ tests/",
    "lint": "eslint src/** tests/** --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "verify": "pnpm prettier && pnpm lint",
    "verify:fix": "pnpm prettier:fix && pnpm lint:fix",
    "test": "jest"
  },
  "dependencies": {
    "@coolscript/syntax": "0.0.7",
    "@coolscript/syntax-concrete": "0.0.0",
    "@coolscript/syntax-scoped": "0.0.0"
  }
}
