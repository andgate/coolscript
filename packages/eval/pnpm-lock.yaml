lockfileVersion: 5.3

specifiers:
  '@coolscript/syntax-concrete': ^0.1.0
  '@coolscript/syntax-scoped': ^0.1.0

devDependencies:
  '@coolscript/syntax-concrete': 0.1.0
  '@coolscript/syntax-scoped': 0.1.0

packages:

  /@coolscript/syntax-concrete/0.1.0:
    resolution: {integrity: sha512-fSNmdIUNRbRWsDYjXM+W3475RU3Kt0A97u84EwHG8u9/oKZgl4zwSSzHSwm0K/poexulULS8NFSmA8IPoXit8g==}
    dev: true

  /@coolscript/syntax-scoped/0.1.0:
    resolution: {integrity: sha512-O86mzY6eY2N8BynJlMxwICecPQPz3/Otuxa5QS8Q68Qrkp8tuvGHEoCh3pkf9Q/6f5sZOiVadAnswtH7jIQBJQ==}
    dev: true
