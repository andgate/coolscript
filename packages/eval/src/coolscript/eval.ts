import {
  Term,
  Value,
  VError,
  TmError,
  VObject,
  TmObject,
  TmValue
} from '@coolscript/syntax'

type EvalEnv = { [key: string]: Term }

export function evaluate(tm: Term): Value {
  const result = reduce(tm)
  if (result.tag != 'TmValue') {
    const msg = `Irreducible expression encountered: ${JSON.stringify(result)}`
    return VError(msg)
  }
  return result.value
}

export function reduce(tm: Term, env: EvalEnv = {}): Term {
  if (!tm) {
    const msg = `Term undefined: ${tm}`
    return TmError(msg)
  }

  switch (tm.tag) {
    case 'TmVar': {
      const t = env[tm.variable]
      if (!t) {
        return TmError(`Undefined variable encountered: "${tm.tag}".`)
      }
      return reduce(t, env)
    }
    case 'TmValue':
      return tm
    case 'TmLam':
      return tm
    case 'TmCall': {
      const f = reduce(tm.call.caller, env)
      const xs = tm.call.args.map((x) => reduce(x, env))

      if (f.tag != 'TmLam') {
        const msg = `Non-function encountered in application head: ${JSON.stringify(
          tm
        )}`
        return TmError(msg)
      }

      if (xs.length == 0) {
        return reduce(f, env)
      }

      const lam = f.lam
      const args = lam.args
      if (xs.length != args.length) {
        return TmError(`Incorrect number of arguments: ${tm}`)
      }

      const env2 = Object.fromEntries(args.map((v, i) => [v, xs[i]]))
      return reduce(lam.body, { ...env, ...env2 })
    }
    case 'TmLet': {
      const defs = Object.fromEntries(
        tm.let.binders.map((b) => [b.variable, b.body])
      )
      return reduce(tm.let.body, { ...env, ...defs })
    }
    case 'TmObject': {
      const obj = {}
      let isValue = true
      for (const [k, t1] of Object.entries(tm.obj)) {
        const t2 = reduce(t1, env)
        isValue &&= t2.tag == 'TmValue'
        obj[k] = t2
      }
      if (isValue) {
        for (const [k, t] of Object.entries(obj)) {
          obj[k] = (t as TmValue).value
        }
        return TmValue(VObject(obj))
      }
      return TmObject(obj)
    }
    case 'TmDo': {
      return tm
    }
    default:
      return TmError(`Unknown term tag encountered "${tm.tag}".`)
  }
}
