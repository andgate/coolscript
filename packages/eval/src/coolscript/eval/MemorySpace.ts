import { Value } from './Value'

export class MemorySpace {
  members: Map<string, Value> = new Map()

  get(id: string): Value {
    return this.members.get(id)
  }

  put(id: string, value: Value) {
    this.members.set(id, value)
  }
}
