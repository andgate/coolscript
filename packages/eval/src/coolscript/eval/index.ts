import { Term } from '@coolscript/syntax-concrete'
import { Value } from './Value'
import { Interpreter } from './Interpreter'

export function evaluate(tm: Term): Value {
  const interpreter = new Interpreter()
  return interpreter.interpret(tm)
}
