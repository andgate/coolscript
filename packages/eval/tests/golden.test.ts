import { examples } from '@coolscript/examples'
import { evaluate } from '@coolscript/eval'

const exampleKeys: string[] = Object.keys(examples)

describe('CoolScript Standard Evaluator Golden Test Suite', () => {
  test.each(exampleKeys)('Evaluating example %p', (exampleKey) => {
    const example = examples[exampleKey]
    const parseResult = evaluate(example)
    expect(parseResult).toMatchSnapshot(exampleKey)
  })
})
