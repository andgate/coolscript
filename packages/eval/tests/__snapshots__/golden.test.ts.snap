// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "bool1": bool1 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 5,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "bool": true,
    "tag": "BooleanValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "bool2": bool2 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 6,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "bool": false,
    "tag": "BooleanValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "call1": call1 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": Scope {
          "enclosingScope": null,
          "symbols": Map {},
        },
        "symbols": Map {
          "f" => Object {
            "id": "f",
            "scope": [Circular],
          },
        },
      },
      "span": Object {
        "column": Object {
          "end": 24,
          "start": 23,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 1,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "example1": example1 1`] = `
Object {
  "value": Object {
    "elements": Array [
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 15,
              "start": 12,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "F",
        "tag": "StringValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 20,
              "start": 17,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "o",
        "tag": "StringValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 25,
              "start": 22,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "o",
        "tag": "StringValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 30,
              "start": 27,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "b",
        "tag": "StringValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 35,
              "start": 32,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "a",
        "tag": "StringValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": Scope {
                "enclosingScope": null,
                "symbols": Map {},
              },
              "symbols": Map {
                "id" => Object {
                  "id": "id",
                  "scope": [Circular],
                },
                "constant" => Object {
                  "id": "constant",
                  "scope": [Circular],
                },
                "createFoo" => Object {
                  "id": "createFoo",
                  "scope": [Circular],
                },
                "bar" => Object {
                  "id": "bar",
                  "scope": [Circular],
                },
              },
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 40,
              "start": 37,
            },
            "line": Object {
              "end": 6,
              "start": 6,
            },
            "type": "Span",
          },
        },
        "str": "r",
        "tag": "StringValue",
      },
    ],
    "tag": "ArrayValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "example2": example2 1`] = `
Object {
  "errors": Array [
    [Error: Parser failed! Error message: Syntax error at line 12 col 18:

      j = s.length -
                   ^
Unexpected keyword token: "-". Instead, I was expecting to see one of the following:

A ";" based on:
    statement_list_tail → _ ● ";" _ statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A newline token based on:
    _ → _ ● %newline _
    index_access_term → bterm ● _ "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A comment token based on:
    _ → _ ● %comment _
    index_access_term → bterm ● _ "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "(" based on:
    call_term → bterm _ ● "(" _ call_arguments _ call_term$ebnf$1 ")"
    bterm →  ● call_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "." based on:
    member_access_term → bterm _ ● "." _ identifier_token
    bterm →  ● member_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "[" based on:
    index_access_term → bterm _ ● "[" _ bterm _ "]"
    bterm →  ● index_access_term
    cterm →  ● bterm
    term →  ● cterm
    assignment_statement → identifier_token _ "=" _ ● term
    statement →  ● assignment_statement
    statement_list_tail → _ ";" _ ● statement
    statement_list$ebnf$1 → statement_list$ebnf$1 ● statement_list_tail
    statement_list → statement ● statement_list$ebnf$1
    block_statement$ebnf$1 →  ● statement_list
    block_statement → "{" _ ● block_statement$ebnf$1 _ block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A "}" based on:
    block_statement → "{" _ block_statement$ebnf$1 _ block_statement$ebnf$2 ● "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
A ";" based on:
    block_statement$ebnf$2$subexpression$1 →  ● ";" _
    block_statement$ebnf$2 → block_statement$ebnf$2 ● block_statement$ebnf$2$subexpression$1
    block_statement → "{" _ block_statement$ebnf$1 _ ● block_statement$ebnf$2 "}"
    do_term → "do" _ ● block_statement
    cterm →  ● do_term
    term →  ● cterm
    lambda_term → "(" _ lambda_arguments _ lambda_term$ebnf$1 ")" _ "=>" _ ● term
    cterm →  ● lambda_term
    term →  ● cterm
    variable_declaration → identifier_token _ "=" _ ● term
    declaration →  ● variable_declaration
    let_declarations_tail → _ ";" _ ● declaration
    let_declarations$ebnf$1 → let_declarations$ebnf$1 ● let_declarations_tail
    let_declarations → declaration ● let_declarations$ebnf$1
    let_term → "let" _ ● let_declarations _ let_term$ebnf$1 "in" _ term
    cterm →  ● let_term
    term →  ● cterm
    file → _ ● term _
    main →  ● file
],
  ],
  "value": null,
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "lambda1": lambda1 1`] = `
Object {
  "value": Object {
    "args": Array [
      "x",
    ],
    "body": Object {
      "ann": Object {
        "scope": Scope {
          "enclosingScope": Scope {
            "enclosingScope": null,
            "symbols": Map {},
          },
          "symbols": Map {
            "x" => Object {
              "id": "x",
              "scope": [Circular],
            },
          },
        },
        "span": Object {
          "column": Object {
            "end": 9,
            "start": 8,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
      },
      "tag": "VariableTerm",
      "variable": "x",
    },
    "tag": "LambdaValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "let1": let1 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": Scope {
          "enclosingScope": null,
          "symbols": Map {},
        },
        "symbols": Map {
          "x" => Object {
            "id": "x",
            "scope": [Circular],
          },
        },
      },
      "span": Object {
        "column": Object {
          "end": 10,
          "start": 9,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 1,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "let2": let2 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": Scope {
          "enclosingScope": null,
          "symbols": Map {},
        },
        "symbols": Map {
          "x" => Object {
            "id": "x",
            "scope": [Circular],
          },
        },
      },
      "span": Object {
        "column": Object {
          "end": 10,
          "start": 9,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 1,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "let3": let3 1`] = `
Object {
  "value": Object {
    "elements": Array [
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": null,
              "symbols": Map {},
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
              "y" => Object {
                "id": "y",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 10,
              "start": 9,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
        },
        "num": 1,
        "tag": "NumberValue",
      },
      Object {
        "ann": Object {
          "scope": Scope {
            "enclosingScope": Scope {
              "enclosingScope": null,
              "symbols": Map {},
            },
            "symbols": Map {
              "x" => Object {
                "id": "x",
                "scope": [Circular],
              },
              "y" => Object {
                "id": "y",
                "scope": [Circular],
              },
            },
          },
          "span": Object {
            "column": Object {
              "end": 17,
              "start": 16,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
        },
        "num": 2,
        "tag": "NumberValue",
      },
    ],
    "tag": "ArrayValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "null1": null1 1`] = `
Object {
  "errors": Array [
    [Error: No parse results were recovered.],
  ],
  "value": null,
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "null2": null2 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 5,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "tag": "NullValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "number1": number1 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 2,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 0,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "number2": number2 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 11,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 1234567890,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "number3": number3 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 22,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "num": 1234567890.0123458,
    "tag": "NumberValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string1": string1 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 3,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "",
    "tag": "StringValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string2": string2 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 3,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "",
    "tag": "StringValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string3": string3 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 15,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "Hello world!",
    "tag": "StringValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string4": string4 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 15,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "Hello World!",
    "tag": "StringValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string5": string5 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 47,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "The quick brown fox jumps over the lazy dog.",
    "tag": "StringValue",
  },
}
`;

exports[`CoolScript Standard Evaluator Golden Test Suite Evaluating example "string6": string6 1`] = `
Object {
  "value": Object {
    "ann": Object {
      "scope": Scope {
        "enclosingScope": null,
        "symbols": Map {},
      },
      "span": Object {
        "column": Object {
          "end": 47,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
    },
    "str": "The quick brown fox jumps over the lazy dog.",
    "tag": "StringValue",
  },
}
`;
